Fix 821: Warn on x === [], x === {}, etc

Rebraws opened the pull request:
Fixes https://github.com/quick-lint/quick-lint-js/issues/821

Now quick-lint-js gives a warning when using strict equality or inequality operators for comparing against an array literal, object literal, class literal, arrow function, or regexp literal.

strager commented on test/test-parse-warning.cpp:
```diff
@@ -327,6 +327,262 @@ TEST_F(test_parse_warning, warn_on_pointless_string_compare_literals) {
+}
+
+TEST_F(test_parse_warning,
+       warn_on_pointless_non_strict_compare_against_class_literal) {
```
Nit: I'd merge all the no-warn test cases together. Maybe call the merged test `non_strict_compare_against_literal_is_not_a_warning`. Then I'd loop over test data. See `test_parse_var.new_style_variables_cannot_be_named_let` in test/test-parse-var.cpp for example.

strager reviewed (approved):
Overall, I think your code looks good. But it's not warning in cases it should warn.

Rebraws commented on src/quick-lint-js/fe/diagnostic-types.h:
```diff
@@ -2197,6 +2197,41 @@
                                                                                \
+  QLJS_DIAG_TYPE(                                                              \
+      diag_pointless_strict_comp_against_array_literal, "E0290",               \
+      diagnostic_severity::warning, { source_code_span span_operator; },       \
```
operator is a C++ keyword, maybe comparison_operator could be better?

strager replied:
Ah, right.

Yeah, `comparison_operator` is better. Or `equals_operator`, which is what we use for `diag_equals_does_not_distribute_over_or` (E0190).

Rebraws commented on docs/errors/E0290.md:
```diff
@@ -0,0 +1,29 @@
+# E0290: using '===' or '!==' against an array literal always produces the same result
```
Yeah, it definitely looks better this way, but when I was fixing this, I was thinking if it could be a good idea to addionally give a warning like `this comparison seems unintentional` when the user uses a non-strict operator. I know that in some cases `x == []` might be true, but I don't think that there's a valid reason to write that (maybe there is, I'm not extremely familiar with js)


So we could have three diagnostics for arrays:
For strict operators (=== and !==):
  - `'{0} []' is always {1}` ('=== []' is always true; '!== []' is always false)
  -  `using '{0}' against an array literal does not compare items`

for non-strict operators (== and !=):
  - `this comparison seems unintentional` 

Do you think it could be a good idea?

strager replied:
Let's not warn on `==` or `!=` for now. That can be a different task.

Rebraws commented:
I think I've addressed all the comments, let me know if anything else needs to be changed or improved

 -  Implemented the two diagnostics for array literals as suggested
 -  Now obj.prop === []  gives a proper warning 
 -  Removed the unnecesary lambda function and removed the overhead in the is_regexp_literal function
 -  Refactor the tests cases to use a for loop
 -  Provided better docs examples and removed the ones without fix suggestions 


I did not provide a warning for `==` and `!=` when compared against arrays literals

strager commented on src/quick-lint-js/fe/diagnostic-types.h:
```diff
@@ -2288,6 +2288,67 @@
+  QLJS_DIAG_TYPE(                                                              \
+      diag_pointless_comp_against_arrow_function, "E0302",                     \
+      diagnostic_severity::warning,                                            \
+      {                                                                        \
+        source_code_span equals_operator;                                      \
+        string8_view condition;                                                \
+      },                                                                       \
+      MESSAGE(QLJS_TRANSLATABLE("using '{0}' against an arrow "                \
+                                "function always returns '{1}'"),              \
+              equals_operator, condition))                                     \
```
Nit: I think `result` or `comparison_result` is a better name than `condition`:
```suggestion
      {                                                                        \
        source_code_span equals_operator;                                      \
        string8_view result;                                                   \
      },                                                                       \
      MESSAGE(QLJS_TRANSLATABLE("using '{0}' against an arrow "                \
                                "function always returns '{1}'"),              \
              equals_operator, result))                                        \
```

strager commented on src/quick-lint-js/fe/parse.cpp:
```diff
@@ -402,6 +402,80 @@ void parser::error_on_invalid_as_const(expression* ast,
+    return op == u8"==="sv || op == u8"!=="sv;
+  };
+  auto is_regexp_literal = [](expression::literal* reg_exp) -> bool {
+    return !(reg_exp->span().begin() == reg_exp->span().end() ||
+             *reg_exp->span().begin() != u8'/');
```
Nit: Avoid calling `span` many times:
```suggestion
    source_code_span span = reg_exp->span();
    return !(span.begin() == span.end() ||
             *span.begin() != u8'/');
```

strager commented on src/quick-lint-js/fe/parse.cpp:
```diff
@@ -402,6 +402,80 @@ void parser::error_on_invalid_as_const(expression* ast,
+        break;
+      }
+    }
+  EXIT_LOOP:;
```
Nit: Instead of `goto`, let's make a `parser` member function.

strager commented on src/quick-lint-js/fe/parse.cpp:
```diff
@@ -402,6 +402,80 @@ void parser::error_on_invalid_as_const(expression* ast,
+    for (expression* child : {ast->child(i), ast->child(i + 1)}) {
+      string8_view condition = get_condition_result(op_span.string_view());
+
+      switch (child->without_paren()->kind()) {
+      case expression_kind::_class:
+        this->diag_reporter_->report(diag_pointless_comp_against_class_literal{
+            .equals_operator = op_span, .condition = condition});
+        goto EXIT_LOOP;
+        break;
+      case expression_kind::array:
+        if (is_strict_operator(op_span.string_view())) {
+          if (child->without_paren()->span().string_view() == u8"[]"sv) {
+            this->diag_reporter_->report(
+                diag_pointless_strict_comp_against_empty_array_literal{
+                    .equals_operator = op_span, .condition = condition});
+          } else {
+            this->diag_reporter_->report(
+                diag_pointless_strict_comp_against_array_literal{
+                    .equals_operator = op_span});
+          }
+          goto EXIT_LOOP;
+        }
+        break;
+      case expression_kind::arrow_function:
+        this->diag_reporter_->report(diag_pointless_comp_against_arrow_function{
+            .equals_operator = op_span, .condition = condition});
+        goto EXIT_LOOP;
+        break;
+      case expression_kind::object:
+        this->diag_reporter_->report(diag_pointless_comp_against_object_literal{
+            .equals_operator = op_span, .condition = condition});
+        goto EXIT_LOOP;
+        break;
+      case expression_kind::literal:
+        if (is_regexp_literal(
+                static_cast<expression::literal*>(child->without_paren()))) {
```
Minor: Avoid multiple calls to `without_paren`:
```suggestion
      child = child->without_paren();
      switch (child->kind()) {
      case expression_kind::_class:
        this->diag_reporter_->report(diag_pointless_comp_against_class_literal{
            .equals_operator = op_span, .condition = condition});
        goto EXIT_LOOP;
        break;
      case expression_kind::array:
        if (is_strict_operator(op_span.string_view())) {
          if (child->span().string_view() == u8"[]"sv) {
            this->diag_reporter_->report(
                diag_pointless_strict_comp_against_empty_array_literal{
                    .equals_operator = op_span, .condition = condition});
          } else {
            this->diag_reporter_->report(
                diag_pointless_strict_comp_against_array_literal{
                    .equals_operator = op_span});
          }
          goto EXIT_LOOP;
        }
        break;
      case expression_kind::arrow_function:
        this->diag_reporter_->report(diag_pointless_comp_against_arrow_function{
            .equals_operator = op_span, .condition = condition});
        goto EXIT_LOOP;
        break;
      case expression_kind::object:
        this->diag_reporter_->report(diag_pointless_comp_against_object_literal{
            .equals_operator = op_span, .condition = condition});
        goto EXIT_LOOP;
        break;
      case expression_kind::literal:
        if (is_regexp_literal(
                static_cast<expression::literal*>(child))) {
```

strager commented on src/quick-lint-js/fe/parse.cpp:
```diff
@@ -402,6 +402,80 @@ void parser::error_on_invalid_as_const(expression* ast,
+        break;
+      case expression_kind::array:
+        if (is_strict_operator(op_span.string_view())) {
+          if (child->without_paren()->span().string_view() == u8"[]"sv) {
```
Must fix: This doesn't work if the array literal has a space or comment inside. Instead of comparing strings, query if the array is any children.
```suggestion
          if (child->without_paren()->child_count() == 0) {
```

strager reviewed (changes requested):
See 'must fix' comment. Otherwise, this looks good to me.

I can deal with the rebasing if you want.

Rebraws commented:
I did the rebase and I've addressed all the comments from the last review, I've also noticed that now the `expression::literal` class  has the `is_regexp` member function, so I deleted the `is_regexp_literal` lambda function and just used `static_cast<expression::literal*>(child)->is_regexp()`

There are some CI errors, but I think they are not related to the PR.
